findserversonnetwork.c

```diff
@@ -171,71 +149,61 @@ void ualds_findserversonnetwork_removeServiceEntries(ualds_browseContext *pBrows
 
 void ualds_findserversonnetwork_socketEventCallback(int* shutdown)
 {
+    fd_set readFDs;
+    struct timeval tv;
+    int fd, status;
+    ualds_genericContext* context;
+    MulticastSocketCallbackStruct* socketCallbackStruct;
+
+
+    /*
+    TODO:
+    This should be combined with zerconf socket callback into one "poll" loop that builds 
+    pollfd or select bitmask from the socket cb list. This would then be driven from main.
+    */
+    
+
+    /* Lock access to shared sdref/fd from server processing thread */
+    OpcUa_Mutex_Lock(g_mutex);
     OpcUa_List_Enter(&g_findServersSocketList);
     OpcUa_List_ResetCurrent(&g_findServersSocketList);
 
-    MulticastSocketCallbackStruct* socketCallbackStruct = OpcUa_Null;
-
     socketCallbackStruct = (MulticastSocketCallbackStruct*)OpcUa_List_GetCurrentElement(&g_findServersSocketList);
     while (socketCallbackStruct)
     {
-        if (*shutdown)
+        context = (ualds_genericContext*)socketCallbackStruct->context;
+        if (*shutdown || context == OpcUa_Null || context->sdRef == OpcUa_Null)
         {
-            // exit ASAP
-            OpcUa_List_Leave(&g_findServersSocketList);
-            return;
+            /* browse or resolve failed or was stopped - remove socket from poll list */
+            OpcUa_Free(socketCallbackStruct);
+            OpcUa_List_DeleteCurrentElement(&g_findServersSocketList);
+            socketCallbackStruct = (MulticastSocketCallbackStruct*)OpcUa_List_GetCurrentElement(&g_findServersSocketList);
+            continue;
         }
 
-        fd_set         readFDs;
-        struct timeval tv;
-
+        fd = DNSServiceRefSockFD(socketCallbackStruct->sdRef);
         FD_ZERO(&readFDs);
-        FD_SET(socketCallbackStruct->fd, &readFDs);
+        FD_SET(fd, &readFDs);
 
         tv.tv_sec = 0;
         tv.tv_usec = 100000;
```

Comment:
Using a constant would make sense, since it is used a couple of times in the file.